modules/gltf/gltf_document.cpp

```diff
@@ -6858,32 +6881,81 @@ Node *GLTFDocument::generate_scene(Ref<GLTFState> state, int32_t p_bake_fps) {
 			_import_animation(state, ap, i, p_bake_fps);
 		}
 	}
-
+	for (Map<GLTFNodeIndex, Node *>::Element *node_i = state->scene_nodes.front(); node_i; node_i = node_i->next()) {
+		Node *node = state->scene_nodes[node_i->key()];
+		ERR_CONTINUE(!node);
+		for (int32_t ext_i = 0; ext_i < document_extensions.size(); ext_i++) {
+			Ref<GLTFDocumentExtension> ext = document_extensions[ext_i];
+			ERR_CONTINUE(ext.is_null());
+			ERR_CONTINUE(!state->json.has("nodes"));
+			Array nodes = state->json["nodes"];
+			ERR_CONTINUE(node_i->key() >= nodes.size());
+			Dictionary node_json = nodes[node_i->key()];
+			Ref<GLTFNode> gltf_node = state->nodes[node_i->key()];
+			err = ext->import_node(state, gltf_node, node_json, node);
+			ERR_CONTINUE(err != OK);
+		}
+	}
 	for (int32_t ext_i = 0; ext_i < document_extensions.size(); ext_i++) {
 		Ref<GLTFDocumentExtension> ext = document_extensions[ext_i];
 		ERR_CONTINUE(ext.is_null());
-		Error err = ext->import_post(this, root);
+		err = ext->import_post(root);
 		ERR_FAIL_COND_V(err != OK, nullptr);
 	}
 	ERR_FAIL_NULL_V(root, nullptr);
 	return root;
 }
 
-Error GLTFDocument::append_from_scene(Node *p_node, Ref<GLTFState> state, uint32_t p_flags, int32_t p_bake_fps) {
-	ERR_FAIL_COND_V(state.is_null(), FAILED);
-	state->use_named_skin_binds =
+Error GLTFDocument::append_from_scene(Node *p_node, Ref<GLTFState> r_state, uint32_t p_flags, int32_t p_bake_fps) {
+	ERR_FAIL_COND_V(r_state.is_null(), FAILED);
+	r_state->use_named_skin_binds =
 			p_flags & EditorSceneFormatImporter::IMPORT_USE_NAMED_SKIN_BINDS;
-
-	_convert_scene_node(state, p_node, -1, -1);
-	if (!state->buffers.size()) {
-		state->buffers.push_back(Vector<uint8_t>());
+	Error err = OK;
+	bool is_mutating = false;
+	for (int32_t ext_i = 0; ext_i < document_extensions.size(); ext_i++) {
+		Ref<GLTFDocumentExtension> ext = document_extensions[ext_i];
+		ERR_CONTINUE(ext.is_null());
+		if (ext->is_changing_original_tree()) {
+			is_mutating = true;
+			break;
+		}
+	}
+	if (is_mutating) {
+		p_node = p_node->duplicate();
+	}
+	for (int32_t ext_i = 0; ext_i < document_extensions.size(); ext_i++) {
+		Ref<GLTFDocumentExtension> ext = document_extensions[ext_i];
+		ERR_CONTINUE(ext.is_null());
+		err = ext->export_preflight(p_node);
+		ERR_FAIL_COND_V(err != OK, FAILED);
+	}
+	_convert_scene_node(r_state, p_node, -1, -1);
+	if (!r_state->buffers.size()) {
+		r_state->buffers.push_back(Vector<uint8_t>());
 	}
 
 	/* STEP 1 CONVERT MESH INSTANCES */
-	_convert_mesh_instances(state);
+	_convert_mesh_instances(r_state);
 
 	/* STEP 2 CREATE SKINS */
-	Error err = _serialize_skins(state);
+	err = _serialize_skins(r_state);
+	ERR_FAIL_COND_V(err != OK, ERR_INVALID_DATA);
```

Comment:
these two steps are also done in `_serialize()`

Reply:
Any progress on this issue?

Reply:
generate_buffer and write_to_filesystem aka export appear to be using export_node correctly.

append_from_scene appears aka import appears to be using export_node correctly.

Reply:
I'll probably delete the generate_buffer and write_to_filesystem's versions of export_node.

Reply:
Ok, it seems fine now.

I don't want to discourage you from having both if there's a reason for it, but there needs to be a way to distinguish the two calls if so, for example the InternalImportCategory.

Reply:
_convert_mesh_instances and _serialize_skins are still being done twice.